design/design.md

```diff
@@ -0,0 +1,223 @@
+which returns the next event that has occurred, blocking until such an event is
+available. The events, in our application, can be one of two things:
+* A request has arrived.
+* A response has been sent.
+
+When `CompletionQueue::Next()` returns a request, the event loop posts a
+coroutine onto the JobQueue, to be invoked later; this coroutine, when invoked, 
+calls the corresponding handler, populates the response, and submits the response for sending.
+Posting to the JobQueue in this way frees up the event loop to process more events
+while requests are being handled. Once grpc sends the response, an event is
+placed on the `CompletionQueue`; when `CompletionQueue::Next()` returns such an
+event, signaling the response has been sent, the event loop cleans up the resources
+associated with this request/response.
+
+While events are consumed from the `CompletionQueue` sequentially, responses are
+populated and sent in parallel. `CompletionQueue` is thread safe. It is possible
+to have multiple event loops running in parallel (one `CompletionQueue` and one
+thread per event loop), though this may not necessarily lead to better
+scalablity, since handlers are called from the JobQueue.
+
+## Classes
+The server code is adapted from the example found
+[here](https://github.com/grpc/grpc/blob/v1.24.0/examples/cpp/helloworld/greeter_async_server.cc)
+
+`CompletionQueue` is a class of the GRPC library. `CompletionQueue` is used to queue events, and provides a method, `CompletionQueue::Next()` to query for any events that have occurred. Events are returned by `CompletionQueue` in the form of a `void *` pointer. In our application, `CompletionQueue` returns pointers to `CallData` objects, which are used to service requests.
+
+Each request is represented by a `CallData` object. `CallData` is an abstract
+class that implements a method `Proceed()` and can be in one of three states:
+CREATE, PROCESS or FINISH. These states correspond to different states in the lifecycle of an rpc call, and `Proceed()` transitions the object from one
+state to the next. `CallData` has three pure virtual methods:
+`doCreate()`, `doProcess()` and `doFinish()`, which are implemented by derived
+classes. For each rpc method, there is a class that derives from `CallData`, and
+implements the pure virtual methods.
+
+```
+class CallData {
+    void Proceed() {
+      if (status_ == CREATE) {
+          status_ = PROCESS;
+          doCreate();
+      } else if (status_ == PROCESS) {
+          status_ = FINISH;
+          doProcess();
+      } else {
+          GPR_ASSERT(status_ == FINISH);
+          doFinish();
+      }
+    }
+
+    virtual void doCreate() = 0;
+    virtual void doProcess() = 0;
+    virtual void doFinish() = 0;
+
+    CompletionQueue* cq_;
+    enum CallStatus { CREATE, PROCESS, FINISH};
+    CallStatus status_;
+}
+```
+
+* `doCreate()`, which is called upon object creation (from the constructor), creates a listener for the
```

Comment:
how does the actual "listening" work here. so when an RPC call is received by the server, the CompletionQueue will send a message to CallData, and it will call doProcess? this might go beyond this scope, but what is the rationale here? why can't completion queue just invoke the function directly. perhaps this is to decouple the actual work from signaling to do work?